File: FirstLastSortedArray.java
public class FirstLastSortedArray {

    public static void main(String[] args)
    {
        int nums[] = {1, 2, 3, 4, 5, 6, 7, 8,8,9,7,10,10,11,11,12};
        int target = 8;
        int result[] = searchRange(nums, target);
        System.out.println("result: " + result[0] + " " + result[1]);
    }
    //Time Complexity  : O(log n)
    //Space Complexity : O(n)
    public static int[] searchRange(int[] nums, int target) {
        // 2 binary searches
        // 1 for finding the starting index
        // 2nd for finding the last index 

        int[] result = new int[2];

        int low = 0;
        int high = nums.length - 1;

        result[0] = -1;
        result[1] = -1;

        int first = binarySearch(nums,low,high,target);
        if(first == -1) return result;

        int second = binarySearchLast(nums,first,high,target);

        result[0] = first;
        result[1] = second;
        
        return result;
    }

    public static int binarySearch(int[] nums,int low, int high, int target)
    {
        while(low <= high) 
        {
            int mid = low + (high-low)/2;

            if(nums[mid] == target)
            {
                if(mid ==0 || nums[mid] > nums[mid-1])
                {
                    return mid;
                }
                else 
                {
                    high = mid - 1;
                }
            }
            else if(nums[mid] < target)
            {
                low = mid + 1; 
            }
            else
            {
               high = mid - 1;
            }
        }

    return -1;

    }

    public static int binarySearchLast(int[] nums,int low, int high, int target)
    {
        while(low <= high) 
        {
            int mid = low + (high-low)/2;

            if(nums[mid] == target)
            {
                if(mid == nums.length-1 || nums[mid] < nums[mid+1])
                {
                    return mid;
                }
                else 
                {
                    low = mid + 1;
                }
            }
            else if(nums[mid] < target)
            {
                low = mid + 1; 
            }
            else
            {
               high = mid - 1;
            }
        }

    return -1;

    }
}

File: First_Last_SortedArray.py
'''
Approach:
if target is equal to mid we need to find the first and last occurance of that element
for that we need to initiate the binary search from left of the array and right of the array.

left binary search will return the first index of the element
right binary search will return the last index of the element.

Binary Search:
Calculate the mid and compare the mid with the target element. if target == nums[mid], then return index of the mid
if the target is less than mid then move the high to mid - 1
else move the low to mid + 1
'''

# Time Complexity : O(logn)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

class Solution:

    def binarySearchLeft(self, nums, l, r, target):
        while (l <= r):
            mid = l + (r - l) // 2
            #when nums[mid] == target, then see if immediate left is less than mid, then mid is first occurance
            if nums[mid] == target:
                if (mid == l or nums[mid] > nums[mid - 1]):
                    return mid
                else:
                    r = mid - 1

            elif nums[mid] < target:
                l = mid + 1

            else:
                r = mid - 1

        return -1

    def binarySearchRight(self, nums, l, r, target):
        while (l <= r):
            mid = l + (r - l) // 2
            # when nums[mid] == target, then see if immediate right is greater than mid, then mid is first occurance
            if nums[mid] == target:
                if (mid == r or nums[mid] < nums[mid + 1]):
                    return mid
                else:
                    l = mid + 1

            elif nums[mid] < target:
                l = mid + 1

            else:
                r = mid - 1
        return -1

    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if not nums or len(nums) == 0:
            return [-1, -1]

        low = 0
        high = len(nums) - 1
        left = self.binarySearchLeft(nums, low, high, target)
        if left == -1:
            return [-1, -1]
        right = self.binarySearchRight(nums, low, high, target)

        return [left, right]




